daml/ContingentClaims/Lifecycle.daml

```diff
@@ -3,83 +3,120 @@
+import DA.Traversable (sequence)
 import Prelude hiding (sequence, mapA)
 
 type C a = Claim Observation Date a
 type F a = ClaimF Observation Date a
 
 deriving instance Eq a => Eq (C a)
 
--- | Returned from a lifecycle operation.
+-- | Returned from a `settle` operation.
 data Result a = Result with
   remaining : C a
     -- ^ the tree after lifecycled branches have been pruned.
   pending : [(Decimal, a)]
     -- ^ quantity/asset pairs requiring settlement.
 
--- | Lifecycle a Claim. Parameters:
--- 1. Function taking key and date, used to lookup an observable on the ledger. `m` will typically be `Update`. It is parametrised so it can be run in a `Script`.
--- 2. Function used when branching in an `Or`. Must return one of the arguments.
--- 3. A set of `Claim`s.
--- 4. A function which, given a date, lifecycles the third argument.
-lifecycle : (Eq a, CanAbort m)
+-- | Collect claims falling due into a list, and return the tree with those nodes pruned.
+-- `m` will typically be `Update`. It is parametrised so it can be run in a `Script`.
+-- The first argument is used to lookup the value of any `Observables`.
+-- Results in a function taking today's date, and returning the pruned tree + pending settlements.
+settle : (Eq a, CanAbort m)
   => (Text -> Date -> m Decimal)
-  -> (C a -> C a -> C a)
   -> C a
   -> Date -> m (Result a)
-lifecycle spot choose
+settle spot
   = runKleisli
   . fmap (uncurry Result)
   . flip evalStateT 1.0
   . runWriterT
-  . apoCataM (pruneZeros') (lifecycle' spot choose)
+  . apoCataM pruneZeros' settleLazily'
+    where
+      settleLazily' = (lift . lift <$> lazily' spot)
+                      >=> (fmap distM (fmap (settle' spot . embed)) . sequence)
 
-type Seed m a = WriterT [(Decimal, a)] (StateT Decimal (Kleisli m Date)) (F a (Either (C a) (C a)))
+-- | HIDE The monad stack we're using.
+type M m a = WriterT [(Decimal, a)] (StateT Decimal (Kleisli m Date)) (F a (Either (C a) (C a)))
+--           \----------------------- monad ------------------------/  \-------- value --------/
 
-lifecycle' : (Eq a, Monad m, CanAbort m)
+-- | HIDE Helper function for sequencing algebras.
+distM : Monad m
+      => Either (C a) (WriterT [(Decimal, a)] (StateT Decimal (Kleisli m Date)) (F a (C a)))
+      -> M m a
+distM (Left c) = pure (Left <$> (project c))
+distM (Right m) = fmap Right <$> m
+
+-- | Evaluate observables and skip branches for which predicates don't hold.
+-- This is useless onl it's own. Composed with other functions it adds laziness.
+lazily' : (Eq a, Monad m)
   => (Text -> Date -> m Decimal)
-  -> (C a -> C a -> C a)
-  -> C a -> Seed m a
+  -> C a -> Kleisli m Date (F a (Either (C a) (C a)))
+lazily' spot (When obs c) = do
+  predicate <- eval spot obs
+  pure if predicate then WhenF obs (Right c) else WhenF obs (Left c)
+lazily' spot (Cond obs c c') = do
+  predicate <- eval spot obs
+  if predicate then lazily' spot c else lazily' spot c'
+lazily' _ other = pure $ Right <$> project other
```

Comment:
> leonidr  13 hours ago
> Thinking about the full lifecycle a bit more deeply, I realized that we can't necessarily lifecycle Give automatically either. It is much more than multiplying the payments inside by -1. Imagine if there's an Or inside that Give. What we've done is changed who gets decide how to proceed with that portion of a claim. We can have the "limited" view of Give where it is effectively Scale -1 but with a different name. But I think that within a Daml framework, the "intended" interpretation of Give would require changing controller rights (assuming both the counterparties are signatories).

That's a great point @leonidr-da . The approach I was considering was to also `short-circuit` on a `Give` node here (or possibly in a **separate algebra**). That means that, composed with `exercise`, you would never be able to exercise an `Or` or `Anytime` below a `Give` node.
I'd then need a way to flip that around (so for example a callable bond gives the issuer the right - not the bearer).
In either case, splitting up the effects into separate functions keeps the handling of the `Scale` factor and the concept of who can exercise entirely separate (and can be tested independently).

Reply:
One interesting thing to consider is whether `Give` should only restrict exercise rights, or also settlement rights - i.e. should `One`'s below `Give` only be settled by the counterparty, or by either party?

This is a liveness problem.

One discussion I had with @GeorgSchneider was whether the custodian (counterparty) should be responsible for _initiating_ settlement in call cases. For instance, in the real-world, you don't need to do anything to receive coupons or dividends in your brokerage account; that's part of the safe-keeping role of the custodian. However, not sure if this changes in a trustless environment. As we're assuming here that all claims are positive, then would it make sense that each counterpart should have the right to trigger settlement of their own claims? But then this has the added inconvenience of requiring the bearer to settle his own dividends!
To further complicate matters, recall that the `settle` (previously `lifecycle`) doesn't actually _do_ the settlement - it merely indicates _what needs to be settled_.

Reply:
> be able to exercise an Or or Anytime below a Give node.

I think it is worse than that. If during the lifecycling of a claim, in the broadest interpretation of `Give`, in a trustless environment, I do not have the _right_ to even lifecycle the claim inside. What if it contains an `Observation` to something that at the moment is off-ledger (to be put on ledger) that the new holder is privy to?

Reply:
I think we need to differentiate lifecycling based on who is performing it. In a custodian/clearing-house case (with delegation from the parties), we can have explicit instructions. Such a party would also have restrictions on what time of assets it is willing to lifecycle.

Reply:
Here's another mind-bender: nested `Give`s. How would you handle this:
```haskell
Give (One "USD" `And` Give (One "EUR"))
```

Reply:
The most recent commit adds a boolean argument to the election parameter of `Lifecycle.exercise`, which indicates who is making the election (`True` = the bearer, resp. counterparty).

https://github.com/digital-asset/contingent-claims/pull/15/files#diff-777017d2d9038baba5d5232cbf1a037d054a2d5b65acec36099bea4385efed71R118

It handles nested cases correctly.

For other operations, the authorization for `Give` is ignored; any party can initiate settlement when it's due, or expire a contract past it's horizon (maturity). These help to ensure liveness; i.e. one party might maliciously to 'do nothing' to block the claim from advancing.